symbolic-debuginfo/src/ppdb.rs

```diff
@@ -136,30 +135,82 @@ impl fmt::Debug for PortablePdbObject<'_> {
     }
 }
 
-/// A debug session for a Portable PDB object.
-///
-/// Currently this session is trivial and returns no files or functions.
-pub struct PortablePdbDebugSession;
+/// Debug session for PortablePdb objects.
+pub struct PortablePdbDebugSession<'data> {
+    ppdb: PortablePdb<'data>,
+}
 
-impl<'session> DebugSession<'session> for PortablePdbDebugSession {
-    type Error = Infallible;
+impl<'data> PortablePdbDebugSession<'data> {
+    /// Returns an iterator over all functions in this debug file.
+    pub fn functions(&self) -> PortablePdbFunctionIterator<'_> {
+        iter::empty()
+    }
 
-    type FunctionIterator = PortablePdbFunctionIterator<'session>;
+    /// Returns an iterator over all source files in this debug file.
+    pub fn files(&self) -> PortablePdbFileIterator<'_> {
+        let size = 1; // TODO self.ppdb.get_documents_count()
+        PortablePdbFileIterator {
+            ppdb: &self.ppdb,
+            row: 0,
+            size: size,
+            names: Vec::with_capacity(size),
+        }
+    }
 
+    /// Looks up a file's source contents by its full canonicalized path.
+    ///
+    /// The given path must be canonicalized.
+    pub fn source_by_path(&self, _path: &str) -> Result<Option<Cow<'_, str>>, FormatError> {
+        Ok(None)
+    }
+}
+
+impl<'data, 'session> DebugSession<'session> for PortablePdbDebugSession<'data> {
+    type Error = FormatError;
+    type FunctionIterator = PortablePdbFunctionIterator<'session>;
     type FileIterator = PortablePdbFileIterator<'session>;
 
     fn functions(&'session self) -> Self::FunctionIterator {
-        iter::empty()
+        self.functions()
     }
 
     fn files(&'session self) -> Self::FileIterator {
-        iter::empty()
+        self.files()
     }
 
-    fn source_by_path(
-        &self,
-        _path: &str,
-    ) -> Result<Option<std::borrow::Cow<'_, str>>, Self::Error> {
-        Ok(None)
+    fn source_by_path(&self, path: &str) -> Result<Option<Cow<'_, str>>, Self::Error> {
+        self.source_by_path(path)
+    }
+}
+
+/// An iterator over source files in a DWARF file.
+pub struct PortablePdbFileIterator<'s> {
+    ppdb: &'s PortablePdb<'s>,
+    row: usize,
+    size: usize,
+    // We need to keep `Document.name` list because FileInfo() references into the string.
+    names: Vec<String>,
+}
+
+impl<'s> Iterator for PortablePdbFileIterator<'s> {
+    type Item = Result<FileEntry<'s>, FormatError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.row >= self.size {
+            return None;
+        }
+
+        let index = self.row;
+        self.row += 1;
+
+        let document = self.ppdb.get_document(index).ok()?;
+        self.names[index] = document.name;
```

Comment:
This doesn't look right. I think you would have to append the name here?

Reply:
The issue here, why I was trying what I did in this original version you've reviewed, was that `FileInfo` required reference - it didn't own the data. PPDB get_document() on the other hand produces a `String`. This was incompatible and due to lifetime restrictions, there was no level in the PPDB API to produce correct FileInfo values whose lifetime could be based on `'data`. Therefore, after discussing with @Swatinem, I had to refactor FileInfo (and FileEntry, although that wasn't strictly necessary in this case) to use `Cow`. This is a breaking change if anyone is constructing `FileInfo/FileEntry` directly or reads its public properties instead of getters.